lib/web/apiserver.go

```diff
@@ -523,10 +523,20 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 
 		// serve Web UI:
 		if strings.HasPrefix(r.URL.Path, "/web/app") {
+			// we will set our etag based on the teleport version
+			etag := teleport.Version
```

Comment:
Having the version alone as the ETag _will_ bite people during development. I can guarantee that people will forget about it, run `make build/teleport` followed by starting the cluster and they will wonder why they don't see changes in their assets or JS code.

I wonder if we shouldn't use [`Last-Modified`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified) instead. Everything under `/web/app` is served from `cfg.StaticFS`, which, if I understand correctly, gets created during compilation based on `webassets`.

Since we know that this fs always includes the `index.html` file (at least I assume so by looking at the contents of those two dirs, `teleport` and `e/teleport`, we can do `cfg.StaticFS.Open` to open the index.html file and then stat if for its last modified time. And preferably cache the result so we only need to get the cache time once (`sync.Once` maybe?).

Because the fs gets created during compilation, it is guaranteed to be different between compilations.

Hell, even better, we could incorporate last modified time into our ETag. This would solve the problem with development and guarantee that the timestamp is different between versions, even if we happen to compile two different production versions at the same exact time.

See also https://www.fastly.com/blog/etags-what-they-are-and-how-to-use-them which mentions ETag vs Last-Modified.

Reply:
If we are using a combination of last modified in our etag, then we shouldn't need to use the weak directive. at least as far as i can tell? the whole point of using last modified is to say "this file definitely hasn't changed if the last modified is the same" and if we are incorporating it into our etag, then it is strong and not weak.

Reply:
`Last-Modified` is only about the modification time though, it doesn't say anything about the contents of the file itself. https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#validators We would be fine with strong ETag if we were indeed looking at the response we're serving.

Tangent incoming: I saw people use ETags in situations where the handler for the request computes the response, then they hash the response and add that as a strong ETag. Here's [an example I found in Go](https://github.com/go-http-utils/etag/blob/513ea8f21eb15b2129968302ce01aea91da19b32/etag.go). I think it works similar to how Rails does this. So in the end this just saves transfer on the network, but not necessarily caches anything on the side of the server. Alas, this is not quite what we want to use an ETag for.

Reply:
Looks like getting `Last Modified` time from embedded files[ isn't and won't be supported ](https://github.com/golang/go/issues/44854#issuecomment-792805927).

However, because this change is really only dev experience (avoid the need to `cmd+shift+R` when not using the local dev server), I think this is a compromise that we can make. The solutions are 
- create our own `http.FileSystem` implementation (or rather, extend it to include a compilation time)
- make a timestamp when the process is started and use that as the "current" last modified. this would mean that the file would redownload after every process restart which would be a little weird for prod, but i'm not sure how often our customers are restarting their proxies
- append some sort of flag in the `make build/teleport` command that will set a compilation time somewhere that can be accessed maybe similar to `teleport.Version`
- maybe create a file in the webassets build that is used for "last modified" data? like a text file that includes a timestamp with the name of "last_modified" or something?

All of these seem to be much more involved than force reloading the browser so I'll write an issue and suggest we avoid larger changes

Reply:
From the approaches you listed, adding a file with last modified date sounds most appealing to me. And good call on creating the issue.